src/gennylib/src/dbcheck.cpp

```diff
@@ -0,0 +1,150 @@
+#include <gennylib/dbcheck.hpp>
+#include <gennylib/v1/Topology.hpp>
+
+// Logic in this file is based on the following two js implementations:
+// https://github.com/10gen/workloads/blob/aeaf42b86bb8f1af9bc6ac90198ac0b4ff32bd14/utils/mongoshell.js#L481
+// https://github.com/mongodb/mongo-perf/blob/bd8901a2e76d2fb13d2a6a313f7a9e1bf6be9c04/util/utils.js#L384-L387
+
+namespace genny {
+
+const int DROPPED_COLLECTION_RETRIES = 1000;
+
+namespace {
+bool clearHealthLog(v1::Topology& topology) {
+    class ClearHealthLogVisitor : public v1::TopologyVisitor {
+    public:
+        void onReplSetMongod(const v1::MongodDescription& desc) override {
+            mongocxx::client client(mongocxx::uri(desc.mongodUri));
+            auto local = client["local"];
+            auto healthLog = local["system.healthlog"];
+            healthLog.drop();
+            BOOST_LOG_TRIVIAL(debug) << "Finished clearing healthLog on node: " << desc.mongodUri;
+            _success = true;
+        };
+
+        bool success() {
+            return _success;
+        }
+
+    private:
+        bool _success = false;
+    };
+
+    ClearHealthLogVisitor clearHealthLogVisitor;
+    topology.accept(clearHealthLogVisitor);
+    return clearHealthLogVisitor.success();
+}
+
+bool waitForDbCheckToFinish(v1::Topology& topology) {
+    class WaitForDbCheckVisitor : public v1::TopologyVisitor {
+    public:
+        void onReplSetMongod(const v1::MongodDescription& desc) override {
+            BOOST_LOG_TRIVIAL(debug) << "Waiting for dbcheck to finish on node: " << desc.mongodUri;
+
+            mongocxx::client client(mongocxx::uri(desc.mongodUri));
+            auto local = client["local"];
+            auto healthLog = local["system.healthlog"];
+            // Dbcheck end health log.
+            auto query = bsoncxx::builder::stream::document{} << "operation"
+                                                              << "dbCheckStop"
+                                                              << bsoncxx::builder::stream::finalize;
+            int count = 0;
+            while (!count) {
+                count = healthLog.count_documents(query.view());
+            }
+
+            BOOST_LOG_TRIVIAL(debug) << "dbcheck finished on node: " << desc.mongodUri;
+            _success = true;
+        };
+
+        bool success() {
+            return _success;
+        }
+
+    private:
+        bool _success = false;
+    };
+
+    WaitForDbCheckVisitor waitForDbCheckVisitor;
+    topology.accept(waitForDbCheckVisitor);
+    return waitForDbCheckVisitor.success();
+}
+
+}  // anonymous namespace
+
+bool dbcheck(mongocxx::pool::entry& client,
+             mongocxx::database& db,
+             std::string& collName,
+             bsoncxx::document::value& dbCheckParam) {
+    /* Only one thread needs to actually do the dbcheck. If
+     * another thread enters the function while a dbcheck
+     * is happening, it waits until the dbcheck finishes
+     * and exits.
```

Comment:
We only support one Thread running this, so I think this can go away and we can avoid some complexity.

Reply:
yes, but it can be called from different actor in the same phase, right ?

Reply:
See my other comment, but I think using this in multiple Actors (i.e. using it in any Actor outside of your DbCheckActor) would be a sign that the Actor(s) involved don't have the right level of granularity.

Actors being single-threaded is how we get away with not having to worry about concurrency in a lot of Actor and Actor-support code. If we have to worry about synchronization across Actors in a way that isn't "Orchestrator-aware" it can lead to very hard to reason-about races or timing bugs since this will compete with the Orchestrator's job of ensuring the cross-Actor state is sane.

If you decide to keep this as a separate header/lib--and if you decide you wanna keep this--then please add commentary that this should never happen because only one DbCheckActor may be instantiated, and this function may only be called from DbCheckActor.